client/src/main/java/org/eclipse/hono/client/impl/AbstractRequestResponseClient.java

```diff
@@ -0,0 +1,207 @@
+
+            receiverTracker.compose(openReceiver -> {
+                this.receiver = openReceiver;
+                ProtonSender sender = con.createSender(String.format(requestResponseAddressTemplate, tenantId));
+                sender
+                        .setQoS(ProtonQoS.AT_LEAST_ONCE)
+                        .openHandler(senderTracker.completer())
+                        .open();
+            }, senderTracker);
+        });
+    }
+
+    /**
+     * Build a Proton message with a provided subject (serving as the operation that shall be invoked).
+     * The message can be extended by arbitrary application properties passed in.
+     * <p>
+     * To enable specific message properties that are not considered here, the method can be overridden by subclasses.
+     *
+     * @param subject The subject system property of the message.
+     * @param appProperties The map containing arbitrary application properties.
+     *                      Maybe null if no application properties are needed.
+     * @return The Proton message constructed from the provided parameters.
+     */
+    protected Message createMessage(final String subject, final Map<String, Object> appProperties) {
+        final Message msg = ProtonHelper.message();
+        final String messageId = createMessageId();
+        msg.setApplicationProperties(new ApplicationProperties(appProperties));
+        msg.setReplyTo(replyToAddress);
+        msg.setMessageId(messageId);
+        msg.setSubject(subject);
+        return msg;
+    }
+
+
+    private R getRequestResponseResult(final Message message) {
+        final String status = MessageHelper.getApplicationProperty(
+                message.getApplicationProperties(),
+                MessageHelper.APP_PROPERTY_STATUS,
+                String.class);
+        final JsonObject payload = MessageHelper.getJsonPayload(message);
+        return getResult(Integer.valueOf(status), payload);
+    }
+
+    protected final void createAndSendRequest(final String action, final JsonObject payload,
+                                        final Handler<AsyncResult<R>> resultHandler) {
+        createAndSendRequest(action,null,payload,resultHandler);
+    }
+
+    protected final void createAndSendRequest(final String action, final Map<String, Object> properties, final JsonObject payload,
+                                      final Handler<AsyncResult<R>> resultHandler) {
+
+        final Message request = createMessage(action, properties);
+        if (payload != null) {
+            request.setContentType("application/json; charset=utf-8");
+            request.setBody(new AmqpValue(payload.encode()));
+        }
+        sendMessage(request, resultHandler);
+    }
+
+    protected final void sendMessage(final Message request, final Handler<AsyncResult<R>> resultHandler) {
+
+        context.runOnContext(req -> {
+            replyMap.put((String) request.getMessageId(), resultHandler);
+            sender.send(request);
```

Comment:
we currently assume that we always have enough credit (which seems fair at the moment) but I wonder if (in a future PR) we should add functionality for waiting on more credits before sending out additional requests when we run out of credit ...